File: code/comfyui/worker/csharp/ComfyWorker/RabbitMqJobQueue.cs
using System.Text;
using System.Text.Json.Nodes;
using System.Threading.Channels;
using RabbitMQ.Client;
using RabbitMQ.Client.Events;

namespace Learn.Comfy.Worker;

/// <summary>
/// The same <see cref="IJobQueue"/> over RabbitMQ: a durable quorum queue of jobs, manual acknowledgements, a prefetch
/// equal to the number of GPUs, and a second queue for dead letters. The message is the job as JSON:
/// {"id": "…", "workflow": {…}}, with the job id as message id.
/// </summary>
public sealed class RabbitMqJobQueue : IJobQueue, IAsyncDisposable
{
    readonly IConnection connection;
    readonly IChannel channel;
    readonly string deadLetterQueue;
    readonly Channel<IDelivery> deliveries = Channel.CreateUnbounded<IDelivery>();

    RabbitMqJobQueue(IConnection connection, IChannel channel, string deadLetterQueue)
    {
        this.connection = connection;
        this.channel = channel;
        this.deadLetterQueue = deadLetterQueue;
    }

    public static async Task<RabbitMqJobQueue> ConnectAsync(Uri broker, string queue, ushort prefetch, CancellationToken cancel)
    {
        var factory = new ConnectionFactory { Uri = broker, ClientProvidedName = $"comfy-worker {Environment.MachineName}" };
        var connection = await factory.CreateConnectionAsync(cancel);
        var channel = await connection.CreateChannelAsync(cancellationToken: cancel);
        var quorum = new Dictionary<string, object?> { ["x-queue-type"] = "quorum" };
        await channel.QueueDeclareAsync(queue, durable: true, exclusive: false, autoDelete: false, arguments: quorum, cancellationToken: cancel);
        await channel.QueueDeclareAsync($"{queue}.dead", durable: true, exclusive: false, autoDelete: false, arguments: quorum, cancellationToken: cancel);
        // Prefetch = GPUs: the broker never hands this worker more jobs than it can run, the rest stay for other workers.
        await channel.BasicQosAsync(0, prefetch, global: false, cancel);

        var result = new RabbitMqJobQueue(connection, channel, $"{queue}.dead");
        var consumer = new AsyncEventingBasicConsumer(channel);
        consumer.ReceivedAsync += async (_, message) =>
        {
            var json = JsonNode.Parse(Encoding.UTF8.GetString(message.Body.Span))!;
            var job = Job.Create(json["id"]!.GetValue<string>(), json["workflow"]!.AsObject());
            await result.deliveries.Writer.WriteAsync(new Delivery(result, job, message.DeliveryTag, message.Body.ToArray()));
        };
        await channel.BasicConsumeAsync(queue, autoAck: false, consumer, cancel);
        return result;
    }

    public static async Task PublishAsync(Uri broker, string queue, Job job)
    {
        var factory = new ConnectionFactory { Uri = broker };
        await using var connection = await factory.CreateConnectionAsync();
        await using var channel = await connection.CreateChannelAsync(new CreateChannelOptions(publisherConfirmationsEnabled: true, publisherConfirmationTrackingEnabled: true));
        await channel.QueueDeclareAsync(queue, durable: true, exclusive: false, autoDelete: false, arguments: new Dictionary<string, object?> { ["x-queue-type"] = "quorum" });
        var body = new JsonObject { ["id"] = job.Id, ["workflow"] = job.Workflow.DeepClone() };
        var properties = new BasicProperties { MessageId = job.Id, Persistent = true, ContentType = "application/json" };
        await channel.BasicPublishAsync("", queue, mandatory: true, properties, Encoding.UTF8.GetBytes(body.ToJsonString()));
    }

    public async ValueTask<IDelivery?> ReceiveAsync(CancellationToken cancel) => await deliveries.Reader.ReadAsync(cancel);

    public async ValueTask DisposeAsync()
    {
        await channel.CloseAsync();
        await connection.CloseAsync();
    }

    sealed class Delivery(RabbitMqJobQueue queue, Job job, ulong tag, byte[] body) : IDelivery
    {
        public Job Job => job;

        public ValueTask AckAsync() => queue.channel.BasicAckAsync(tag, multiple: false);

        public ValueTask RequeueAsync() => queue.channel.BasicNackAsync(tag, multiple: false, requeue: true);

        // Published to the dead-letter queue with the reason in a header, then acked. A crash between the two
        // leaves the job in both queues: a duplicate dead letter, never a lost one.
        public async ValueTask DeadLetterAsync(string reason)
        {
            var properties = new BasicProperties
            {
                MessageId = job.Id,
                Persistent = true,
                ContentType = "application/json",
                Headers = new Dictionary<string, object?> { ["x-worker-reason"] = reason },
            };
            await queue.channel.BasicPublishAsync("", queue.deadLetterQueue, mandatory: false, properties, body);
            await queue.channel.BasicAckAsync(tag, multiple: false);
        }
    }
}

File: code/comfyui/worker/java/src/main/java/dev/learn/comfy/worker/RabbitMqJobQueue.java
package dev.learn.comfy.worker;

import java.io.IOException;
import java.net.URI;
import java.nio.charset.StandardCharsets;
import java.time.Duration;
import java.util.Map;
import java.util.concurrent.LinkedBlockingQueue;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.TimeoutException;

import com.rabbitmq.client.AMQP;
import com.rabbitmq.client.Channel;
import com.rabbitmq.client.Connection;
import com.rabbitmq.client.ConnectionFactory;
import com.rabbitmq.client.MessageProperties;

import dev.learn.comfy.worker.Jobs.Delivery;
import dev.learn.comfy.worker.Jobs.Job;
import dev.learn.comfy.worker.Jobs.JobQueue;
import tools.jackson.databind.JsonNode;
import tools.jackson.databind.node.ObjectNode;

/**
 * The same {@link JobQueue} over RabbitMQ: a durable quorum queue of jobs, manual acknowledgements, a prefetch equal to
 * the number of GPUs, and a second queue for dead letters. The message is {"id": "…", "workflow": {…}}.
 */
public final class RabbitMqJobQueue implements JobQueue, AutoCloseable {
    private static final Map<String, Object> QUORUM = Map.of("x-queue-type", "quorum");

    private final Connection connection;
    private final Channel channel;
    private final String deadLetterQueue;
    private final LinkedBlockingQueue<Delivery> deliveries = new LinkedBlockingQueue<>();

    private RabbitMqJobQueue(Connection connection, Channel channel, String deadLetterQueue) {
        this.connection = connection;
        this.channel = channel;
        this.deadLetterQueue = deadLetterQueue;
    }

    public static RabbitMqJobQueue connect(URI broker, String queue, int prefetch) throws Exception {
        var factory = new ConnectionFactory();
        factory.setUri(broker);
        Connection connection = factory.newConnection("comfy-worker");
        Channel channel = connection.createChannel();
        channel.queueDeclare(queue, true, false, false, QUORUM);
        channel.queueDeclare(queue + ".dead", true, false, false, QUORUM);
        // Prefetch = GPUs: the broker never hands this worker more jobs than it can run.
        channel.basicQos(prefetch);
        var result = new RabbitMqJobQueue(connection, channel, queue + ".dead");
        channel.basicConsume(queue, false, (tag, message) -> {
            JsonNode json = ComfyInstance.JSON.readTree(new String(message.getBody(), StandardCharsets.UTF_8));
            Job job = Job.create(json.path("id").asString(), (ObjectNode) json.path("workflow"));
            result.deliveries.add(result.new RabbitDelivery(job, message.getEnvelope().getDeliveryTag(), message.getBody()));
        }, tag -> { });
        return result;
    }

    public static void publish(URI broker, String queue, Job job) throws Exception {
        var factory = new ConnectionFactory();
        factory.setUri(broker);
        try (Connection connection = factory.newConnection(); Channel channel = connection.createChannel()) {
            channel.queueDeclare(queue, true, false, false, QUORUM);
            channel.confirmSelect();
            ObjectNode body = ComfyInstance.JSON.createObjectNode().put("id", job.id());
            body.set("workflow", job.workflow().deepCopy());
            AMQP.BasicProperties properties = MessageProperties.PERSISTENT_BASIC.builder().messageId(job.id()).contentType("application/json").build();
            channel.basicPublish("", queue, true, properties, body.toString().getBytes(StandardCharsets.UTF_8));
            channel.waitForConfirmsOrDie(10_000);
        }
    }

    @Override
    public Delivery receive(Duration timeout) throws InterruptedException {
        return deliveries.poll(timeout.toMillis(), TimeUnit.MILLISECONDS);
    }

    @Override
    public boolean isDrained() {
        return false; // a broker queue never ends: the worker stops on a signal
    }

    @Override
    public void close() throws IOException, TimeoutException {
        channel.close();
        connection.close();
    }

    // A Channel is not meant to be shared between threads for publishing: the worker's consumer threads
    // serialize their acknowledgements and publications on it.
    private final class RabbitDelivery implements Delivery {
        private final Job job;
        private final long tag;
        private final byte[] body;

        RabbitDelivery(Job job, long tag, byte[] body) {
            this.job = job;
            this.tag = tag;
            this.body = body;
        }

        @Override
        public Job job() {
            return job;
        }

        @Override
        public void ack() throws IOException {
            synchronized (channel) {
                channel.basicAck(tag, false);
            }
        }

        @Override
        public void requeue() throws IOException {
            synchronized (channel) {
                channel.basicNack(tag, false, true);
            }
        }

        /** Published to the dead-letter queue with the reason in a header, then acked. */
        @Override
        public void deadLetter(String reason) throws IOException {
            AMQP.BasicProperties properties = MessageProperties.PERSISTENT_BASIC.builder()
                    .messageId(job.id()).contentType("application/json").headers(Map.of("x-worker-reason", reason)).build();
            synchronized (channel) {
                channel.basicPublish("", deadLetterQueue, properties, body);
                channel.basicAck(tag, false);
            }
        }
    }
}
